plugin/pkg/admission/eventratelimit/admission_test.go

```diff
@@ -0,0 +1,504 @@
+			name:                  "source host should be included in source+object key",
+			serverBurst:           100,
+			sourceObjectBurst:     1,
+			sourceObjectCacheSize: 10,
+			requests: createSourceObjectKeyInclusionRequests(func(label string) *api.Event {
+				return &api.Event{Source: api.EventSource{Host: label}}
+			}),
+		},
+		{
+			name:                  "involved object kind should be included in source+object key",
+			serverBurst:           100,
+			sourceObjectBurst:     1,
+			sourceObjectCacheSize: 10,
+			requests: createSourceObjectKeyInclusionRequests(func(label string) *api.Event {
+				return &api.Event{InvolvedObject: api.ObjectReference{Kind: label}}
+			}),
+		},
+		{
+			name:                  "involved object namespace should be included in source+object key",
+			serverBurst:           100,
+			sourceObjectBurst:     1,
+			sourceObjectCacheSize: 10,
+			requests: createSourceObjectKeyInclusionRequests(func(label string) *api.Event {
+				return &api.Event{InvolvedObject: api.ObjectReference{Namespace: label}}
+			}),
+		},
+		{
+			name:                  "involved object name should be included in source+object key",
+			serverBurst:           100,
+			sourceObjectBurst:     1,
+			sourceObjectCacheSize: 10,
+			requests: createSourceObjectKeyInclusionRequests(func(label string) *api.Event {
+				return &api.Event{InvolvedObject: api.ObjectReference{Name: label}}
+			}),
+		},
+		{
+			name:                  "involved object UID should be included in source+object key",
+			serverBurst:           100,
+			sourceObjectBurst:     1,
+			sourceObjectCacheSize: 10,
+			requests: createSourceObjectKeyInclusionRequests(func(label string) *api.Event {
+				return &api.Event{InvolvedObject: api.ObjectReference{UID: types.UID(label)}}
+			}),
+		},
+		{
+			name:                  "involved object APIVersion should be included in source+object key",
+			serverBurst:           100,
+			sourceObjectBurst:     1,
+			sourceObjectCacheSize: 10,
+			requests: createSourceObjectKeyInclusionRequests(func(label string) *api.Event {
+				return &api.Event{InvolvedObject: api.ObjectReference{APIVersion: label}}
+			}),
+		},
+		{
+			name:          "event blocked by user limits",
+			userBurst:     3,
+			userCacheSize: 10,
+			requests: []request{
+				newEventRequest().withUser("A"),
+				newEventRequest().withUser("A"),
+				newEventRequest().withUser("A"),
+				newEventRequest().withUser("A").blocked(),
+			},
+		},
+		{
+			name: "event from other user not blocked",
+			requests: []request{
+				newEventRequest().withUser("A"),
+				newEventRequest().withUser("A"),
+				newEventRequest().withUser("A"),
+				newEventRequest().withUser("B"),
+			},
+		},
+		{
+			name: "events from other user should not count against limit",
+			requests: []request{
+				newEventRequest().withUser("A"),
+				newEventRequest().withUser("A"),
+				newEventRequest().withUser("B"),
+				newEventRequest().withUser("A"),
+			},
+		},
+	}
+
+TESTCASESLOOP:
```

Comment:
This looks off.  https://blog.golang.org/subtests perhaps?

Reply:
That is convenient. I had not seen that anywhere yet.